### PR TITLE
Rapid Light Dispenser 

### PR DESCRIPTION
**Pull request by Ironnhawk**
## About The Pull Request

Adds the Rapid Light Dispenser for Cyborgs. May need balance to make it as valuable as the one the station can unlock. 

## How This Contributes To The Skyrat Roleplay Experience

Makes it far easier for engineering borgs to light areas and replace lights that have been destroyed. 

## Changelog

:cl:
add: Added new mechanics or gameplay changes
add: Added more things

/:cl:

### CODE REVIEW

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 1]
```diff
@@ -48,3 +48,36 @@
 		tool_behaviour = TOOL_SCREWDRIVER
 		to_chat(user, span_notice("You attach the screw bit to [src]."))
 		icon_state = "drill_screw_cyborg"
+
+// Rapid Light Dispenser 
+
+/obj/item/construction/rld/borg
+	no_ammo_message = "<span class='warning'>Insufficient charge.</span>"
+	desc = "A device used to rapidly build walls and floors."
+	var/energyfactor = 72
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm:58]: code doc this

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 2]
```diff
@@ -48,3 +48,36 @@
...
+
+/obj/item/construction/rld/borg/useResource(amount, mob/user)
+	if(!iscyborg(user))
+		return 0
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: true/false defines

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 3]
```diff
@@ -48,3 +48,36 @@
...
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		if(user)
+			to_chat(user, no_ammo_message)
+		return 0
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: ditto

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 4]
```diff
@@ -48,3 +48,36 @@
...
+	no_ammo_message = "<span class='warning'>Insufficient charge.</span>"
+	desc = "A device used to rapidly build walls and floors."
+	var/energyfactor = 72
+
+/obj/item/construction/rld/borg/useResource(amount, mob/user)
+	if(!iscyborg(user))
+		return 0
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		if(user)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: you already check for user on line 60, no need to check again

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 5]
```diff
@@ -48,3 +48,36 @@
...
+			to_chat(user, no_ammo_message)
+		return 0
+	. = borgy.cell.use(amount * energyfactor) //borgs get 1.3x the use of their RCDs
+	if(!. && user)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm:65]: ditto

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 6]
```diff
@@ -48,3 +48,36 @@
...
+		to_chat(user, no_ammo_message)
+	return .
+
+/obj/item/construction/rld/borg/checkResource(amount, mob/user)
+	if(!iscyborg(user))
+		return 0
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: true/false defines

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 7]
```diff
@@ -48,3 +48,36 @@
...
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		if(user)
+			to_chat(user, no_ammo_message)
+		return 0
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: ditto

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 8]
```diff
@@ -48,3 +48,36 @@
...
+	. = borgy.cell.charge >= (amount * energyfactor)
+	if(!. && user)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: don't need to user check

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 9]
```diff
@@ -48,3 +48,36 @@
...
+		to_chat(user, no_ammo_message)
+	return .
+
+/obj/item/construction/rld/borg/checkResource(amount, mob/user)
+	if(!iscyborg(user))
+		return 0
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		if(user)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]: same here

File: modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm [hunk 10]
```diff
@@ -52,3 +52,11 @@
 	)
 	category = list("Cyborg Upgrade Modules")
 
+/datum/design/borg_upgrade_rapidlightplacer
+	name = "Cyborg Upgrade (Rapid Light Dispensor)"
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm]:
```suggestion
	name = "Cyborg Upgrade (Rapid Light Dispenser)"
```

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm [hunk 11]
```diff
@@ -244,3 +244,33 @@
 		var/obj/item/inducer/cyborg/inducer = locate() in target_robot.model
 		if (inducer)
 			target_robot.model.remove_module(inducer, TRUE)
+
+// Rapid Light Dispensor
+
+/obj/item/borg/upgrade/rapidlightplacer
+	name = "engineering cyborg RLD upgrade"
+	desc = "An inducer device for the engineering cyborg."
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/robot_upgrades.dm]: not an inducer

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm [hunk 12]
```diff
@@ -244,3 +244,33 @@
...
+	icon_state = "cyborg_upgrade3"
+	require_model = TRUE
+	model_type = list(/obj/item/robot_model/engineering, /obj/item/robot_model/saboteur)
+	model_flags = BORG_MODEL_ENGINEERING
+
+/obj/item/borg/upgrade/rapidlightplacer/action(mob/living/silicon/robot/target_robot, user = usr)
+	. = ..()
+	if(.)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/robot_upgrades.dm:260]: early return

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm [hunk 13]
```diff
@@ -244,3 +244,33 @@
...
+			to_chat(user, span_warning("This unit is already equipped with an inducer module!"))
+			return FALSE
+
+		rld = new(target_robot.model)
+		target_robot.model.basic_modules += rld
+		target_robot.model.add_module(rld, FALSE, TRUE)
+
+/obj/item/borg/upgrade/rapidlightplacer/deactivate(mob/living/silicon/robot/target_robot, user = usr)
+	. = ..()
+	if (.)
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/robot_upgrades.dm:273]: ditto

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm [hunk 14]
```diff
@@ -244,3 +244,33 @@
...
+	model_type = list(/obj/item/robot_model/engineering, /obj/item/robot_model/saboteur)
+	model_flags = BORG_MODEL_ENGINEERING
+
+/obj/item/borg/upgrade/rapidlightplacer/action(mob/living/silicon/robot/target_robot, user = usr)
+	. = ..()
+	if(.)
+
+		var/obj/item/construction/rld/borg/rld = locate() in target_robot
+		if(rld)
+			to_chat(user, span_warning("This unit is already equipped with an inducer module!"))
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/robot_upgrades.dm]: not an inducer

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm (hunk 4 above)
Comment (Ironnhawk) [reply modular_skyrat/modules/cyborg/code/borgpowertools.dm]: Think i replaced that check properly

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm (hunk 12 above)
Comment (Ironnhawk) [reply modular_skyrat/modules/cyborg/code/robot_upgrades.dm:260]: Not sure what you mean by this, this is the same code as the older RCD Modules. And I simply am unsure what you are specifying.

Comment (tf-4) [reply modular_skyrat/modules/cyborg/code/robot_upgrades.dm:260]: He means instead of:
```
if(x)
	y(z)
```
do
```
if(!x)
	return
y(z)
```

**Comment by Jolly-66** [comment]
I'd recommend maybe trying to take a stab at porting this upstream? Worth a shot

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 15]
```diff
@@ -48,3 +48,35 @@
 		tool_behaviour = TOOL_SCREWDRIVER
 		to_chat(user, span_notice("You attach the screw bit to [src]."))
 		icon_state = "drill_screw_cyborg"
+
+// Rapid Light Dispenser 
+
+// Sets the energy cost to use the RLD - Fail Message when out of energy. 
+/obj/item/construction/rld/borg
+	no_ammo_message = "<span class='warning'>Insufficient charge.</span>"
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
	no_ammo_message = "Insufficient charge."
```

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 16]
```diff
@@ -48,3 +48,35 @@
...
+	desc = "A device used to rapidly build walls and floors."
+	var/energyfactor = 72
+
+/obj/item/construction/rld/borg/useResource(amount, mob/user)
+	if(!iscyborg(user))
+		return FALSE
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		to_chat(user, no_ammo_message)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
		user.balloon_alert(user, no_ammo_message)
```

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 17]
```diff
@@ -48,3 +48,35 @@
...
+		return FALSE
+	. = borgy.cell.use(amount * energyfactor) //borgs get 1.3x the use of their RCDs
+	if(!. && user)
+		to_chat(user, no_ammo_message)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
		user.balloon_alert(user, no_ammo_message)
```

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 18]
```diff
@@ -48,3 +48,35 @@
...
+	return .
+
+/obj/item/construction/rld/borg/checkResource(amount, mob/user)
+	if(!iscyborg(user))
+		return FALSE
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		to_chat(user, no_ammo_message)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
		user.balloon_alert(user, no_ammo_message)
```

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 19]
```diff
@@ -48,3 +48,35 @@
...
+		return FALSE
+	. = borgy.cell.charge >= (amount * energyfactor)
+	if(!.)
+		to_chat(user, no_ammo_message)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
		user.balloon_alert(user, no_ammo_message)
```

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 20]
```diff
@@ -48,3 +48,35 @@
...
+	return .
+
+/obj/item/construction/rld/borg/checkResource(amount, mob/user)
+	if(!iscyborg(user))
+		return FALSE
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		to_chat(user, no_ammo_message)
+		return FALSE
+	. = borgy.cell.charge >= (amount * energyfactor)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
	. = borgy.cell.charge >= (amount * energyfactor)
```
Same thing here

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm [hunk 21]
```diff
@@ -48,3 +48,35 @@
...
+	var/energyfactor = 72
+
+/obj/item/construction/rld/borg/useResource(amount, mob/user)
+	if(!iscyborg(user))
+		return FALSE
+	var/mob/living/silicon/robot/borgy = user
+	if(!borgy.cell)
+		to_chat(user, no_ammo_message)
+		return FALSE
+	. = borgy.cell.use(amount * energyfactor) //borgs get 1.3x the use of their RCDs
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/borgpowertools.dm]:
```suggestion
	. = borgy.cell?.use(amount * energyfactor) //borgs get 1.3x the use of their RCDs
```
Could we not just check for a cell here?

File: modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm [hunk 22]
```diff
@@ -52,3 +52,11 @@
 	)
 	category = list("Cyborg Upgrade Modules")
 
+/datum/design/borg_upgrade_rapidlightplacer
+	name = "Cyborg Upgrade (Rapid Light Dispenser)"
+	id = "borg_upgrade_rapidlightplacer"
+	build_type = MECHFAB
+	build_path = /obj/item/borg/upgrade/rapidlightplacer
+	materials = list(/datum/material/iron = 14500, /datum/material/glass = 7500, /datum/material/silver = 6000, /datum/material/gold = 1500,  /datum/material/diamond = 200, /datum/material/titanium = 8000, /datum/material/plasma = 2000)
```
Comment (tf-4) [inline modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm]:
```suggestion
	materials = list(
		/datum/material/iron = 14500,
		/datum/material/glass = 7500,
		/datum/material/silver = 6000,
		/datum/material/gold = 1500, 
		/datum/material/diamond = 200,
		/datum/material/titanium = 8000,
		/datum/material/plasma = 2000,
	)
```

**Review by tf-4 — CHANGES_REQUESTED** [review]

**Comment by Gandalf2k15** [comment]
CL

**Comment by YakumoChen** [comment]
powercreep

**Comment by ghost** [comment]
I agree this might be useful for upstream, I thiiiink the janiborgs have it there, if not engi/janiborg would find it useful.
(Also idk if my github just didnt update but dont forget to change the items description from walsl/floors to lights)

File: modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm [hunk 23]
```diff
@@ -52,3 +52,19 @@
...
+	materials = list(
+		/datum/material/iron = 14500,
+		/datum/material/glass = 7500,
+		/datum/material/silver = 6000,
+		/datum/material/gold = 1500, 
+		/datum/material/diamond = 200,
+		/datum/material/titanium = 8000,
+		/datum/material/plasma = 2000,
+	)
+	construction_time = 80
```
Comment (Zonespace27) [inline modular_skyrat/modules/cyborg/code/mechafabricator_designs.dm:69]: second defines

File: modular_skyrat/modules/cyborg/code/borgpowertools.dm (hunk 1 above)
Comment (Zonespace27) [reply modular_skyrat/modules/cyborg/code/borgpowertools.dm:58]: this wasn't resolved

File: modular_skyrat/modules/cyborg/code/robot_upgrades.dm (hunk 13 above)
Comment (Zonespace27) [reply modular_skyrat/modules/cyborg/code/robot_upgrades.dm:273]: early return still needed

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Review by tf-4 — CHANGES_REQUESTED** [review]
Changelog

**Pull request closed without merging**